### PR TITLE
Plain Text Code Generation

### PR DESCRIPTION
**Pull request by cacay**

### CODE REVIEW

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/PlainTextCodeGenerator.kt [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+
+                            // remove events where [receivingHost] is the sender of the data
+                            event.send.host != receivingHost
+                    }
+                    // of events matching above criteria, get set of data receivers
+                    .map { event -> event.recv.host }
+                    // remove [receivingHost] from the set of hosts with whom [receivingHost] needs to
+                    // check for equivocation
+                    .filter { host -> host != receivingHost }
+                    .toSet()
```
Comment (cacay) [inline compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/PlainTextCodeGenerator.kt:420]: You should sort this list for deterministic code generation. Also, do you need `.toSet`? If you just iterate through the whole thing, keep it a list.

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/PlainTextCodeGenerator.kt [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+                receiveBuilder.addStatement(
+                    "var %N = %L",
+                    receiveTmp,
+                    codeGeneratorContext.receive(
+                        typeTranslator(typeAnalysis.type(sender)),
+                        hostsToCheckWith.first()
+                    )
+                )
+                receiveBuilder.beginControlFlow(
+                    "if(%N != %N)",
```
Comment (cacay) [inline compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/PlainTextCodeGenerator.kt:438]: Space after `if`. There are a few places with this issue. Find & Replace

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/AbstractCodeGenerator.kt [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    abstract fun declaration(protocol: Protocol, stmt: DeclarationNode): CodeBlock
+
+    abstract fun update(protocol: Protocol, stmt: UpdateNode): CodeBlock
+
+    abstract fun outParameterInitialization(protocol: Protocol, stmt: OutParameterInitializationNode): CodeBlock
+
+    abstract fun output(protocol: Protocol, stmt: OutputNode): CodeBlock
+}
+
+abstract class SingleProtocolCodeGenerator : AbstractCodeGenerator() {
```
Comment (cacay) [inline compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/AbstractCodeGenerator.kt:44]: I don't think we will ever use this.

**Review by cacay — COMMENTED** [review]
Mostly small comments, with a couple ideas for medium size restructuring.

File: cli/src/main/kotlin/edu/cornell/cs/apl/viaduct/cli/Compile.kt [hunk 4]
```diff
@@ -155,7 +166,20 @@ class Compile : CliktCommand(help = "Compile ideal protocol to secure distribute
         )
         val postprocessedProgram = postprocessor.postprocess(annotatedProgram)
 
-        output.println(postprocessedProgram)
+        if (compileKotlin) {
+
+            // TODO - figure out best way to let code generators know which protocols it is responsible for
```
Comment (rolph-recto) [inline cli/src/main/kotlin/edu/cornell/cs/apl/viaduct/cli/Compile.kt:171]: The `ProtocolInterpreter` interface has a member `availableProtocols` that define the protocols it is supposed to interpret; you can probably do the same thing here

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/AbstractCodeGenerator.kt (hunk 3 above)
Comment (rolph-recto) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/AbstractCodeGenerator.kt:44]: This mirrors `SingleProtocolInterpreter` which is used by ZKP and commitment back ends

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.OutParameterArgumentNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.ProgramNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.ReadNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.SimpleStatementNode
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.StatementNode
+import edu.cornell.cs.apl.viaduct.util.FreshNameGenerator
+
+class BackendCodeGenerator(
+    private val program: ProgramNode,
+    codeGenerators: List<(context: CodeGeneratorContext) -> CodeGenerator>,
```
[the remark this answers is not in the document]
Comment (rolph-recto) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt:42]: This mirrors `ViaductBackend`---maybe pass a `ProtocolComposer` object as arg to `BackendCodeGenerator` constructor?

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+        // create switch statement in main method so program can be run on any host
+        mainFunctionBuilder.beginControlFlow("when(host)")
+
+        // add a global host object for each host
+        for (host: Host in this.program.hosts) {
+            fileBuilder.addProperty(
+                PropertySpec.builder(host.name, Host::class)
+                    .initializer(
+                        CodeBlock.of(
+                            "Host(%S)",
```
[the remark this answers is not in the document]
Comment (ianwilkietomasik) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt:106]: Is this what you mean?
                          "%N(%S)",
                            Host::class,
                            "Host"

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+        private val receiveMember = MemberName(Runtime::class.java.packageName, "receive")
+        private val sendMember = MemberName(Runtime::class.java.packageName, "send")
+
+        val freshNameGenerator: FreshNameGenerator = FreshNameGenerator().apply {
+            val initNames: MutableSet<String> = mutableSetOf("runtime")
+            initNames += program.hosts.map { host -> host.toString() }.toSet().toMutableSet()
+        }
+
+        override fun kotlinName(sourceName: Temporary, protocol: Protocol): String =
+            tempMap.getOrPut(Pair(sourceName, protocol)) { freshNameGenerator.getFreshName((sourceName.name).filter { it.isLetterOrDigit() }) }
```
[the remark this answers is not in the document]
Comment (ianwilkietomasik) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt:301]: Without the filter, temps are generated as $tmp_1, which breaks everything. Should I still remove the $ signs?

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+                    if (protocolAnalysis.participatingHosts(reader!!).contains(host)) {
+                        val protocolCodeGenerator = codeGeneratorMap[readerProtocol]
+                            ?: throw CodeGenerationError("no code generator for protocol ${protocol.asDocument.print()}")
+                        hostFunctionBuilder.addCode("%L", protocolCodeGenerator.receive(host, stmt, protocol, readerProtocol!!, events!!))
+                    }
+                }
+            }
+
+            is SimpleStatementNode -> {
+                if (protocolAnalysis.participatingHosts(stmt).contains(host)) {
```
[the remark this answers is not in the document]
Comment (ianwilkietomasik) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt:189]: For BlockNodes and FunctionCallNodes, these checks don't happen, so adding a global check might break things? Is is the case that all hosts participate in all blockNodes and FunctionCallNodes? If so, then I can factor out this check.

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+            // pass runtime object to [host]'s function
+            hostFunctionBuilder.addParameter("runtime", runtimeClassName)
+
+            // generate code for [host]'s role in [this.program]
+            generate(hostFunctionBuilder, nameAnalysis.enclosingFunctionName(mainBody), mainBody, host)
+            fileBuilder.addFunction(hostFunctionBuilder.build())
+
+            // update switch statement in main method to have an option for [host]
+            mainFunctionBuilder.beginControlFlow("%N ->", host.name)
```
[the remark this answers is not in the document]
Comment (ianwilkietomasik) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt:128]: I managed to get rid of brackets, but adding alice1(runtime) as a new statement automatically puts it on a new line. Not sure if there is a way to fix this.

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (cacay) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt]: You can drop only the first letter, but you should do that _before_ asking for a fresh name. I.e.,
```kotlin
freshNameGenerator.getFreshName(sourceName.name.drop(1))
```

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (cacay) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt]: Do it in one go?
```kotlin
mainFunctionBuilder.addStatement("%N -> %N(%N)", host.name, hostFunName, "runtime")
```

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (cacay) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt:186]: `participatingHosts` gives you a list of all hosts that need to execute that statement. If the current host is not part of this set, there is nothing to execute. This holds for blocks and function calls.

For example, `participatingHosts(BlockNode)` is simply the union of the `participatingHost` of all its children. If host is not in this set, then it is not involved in the execution of any of its children (by definition of union). So you will just go through the for loop generating nothing for the children.

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (cacay) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt]: I meant `%N(%S), Host::class, …`

[the remark this answers is not in the document]
Comment (ianwilkietomasik) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt]: Causes: java.lang.IllegalArgumentException: expected name but was class edu.cornell.cs.apl.viaduct.syntax.Host

[the remark this answers is not in the document]
Comment (cacay) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt]: Sorry, I meant `%T`.

**Pull request merged**